ic-helpers/src/ledger/principal_ext.rs

```diff
@@ -154,20 +154,28 @@ impl LedgerPrincipalExt for Principal {
         result.map(|v| v.0).map_err(|e| format!("{:?}", e))
     }
 
+    #[allow(unused_variables)]
     async fn get_balance(
         &self,
         of: Principal,
         sub_account: Option<Subaccount>,
     ) -> Result<u64, String> {
-        let account = AccountIdentifier::new(of.into(), sub_account);
-        let args = BinaryAccountBalanceArgs {
-            account: account.to_address(),
-        };
-        let result = ic_cdk::call::<_, (Tokens,)>(*self, "account_balance", (args,))
-            .await
-            .map_err(|e| e.1)?
-            .0;
-        Ok(result.get_e8s())
+        #[cfg(target_family = "wasm")]
+        {
+            let account = AccountIdentifier::new(of.into(), sub_account);
+            let args = BinaryAccountBalanceArgs {
+                account: account.to_address(),
+            };
+            let result = ic_cdk::call::<_, (Tokens,)>(*self, "account_balance", (args,))
+                .await
+                .map_err(|e| e.1)?
+                .0;
+            Ok(result.get_e8s())
+        }
+        #[cfg(not(target_family = "wasm"))]
+        {
+            Ok(10u64.pow(15))
+        }
```

Comment:
This behavior would be very unexpected for anyone using this code. It also make it impossible to test cases when the balance is not sufficient etc.

When the factory creates a canister, there's logic there not to check the balance or consume ICP in case the caller of the request is the factory controller. If you use that logic for your tests, this hack will not be needed.

To mock the calller principal, you can use ic-kit library.

Reply:
Yeah, I didn't bother with it much because the ledger canister will be rewritten as a canister anyway. It was problematic to rewrite it into `ic-kit` call because of 

```
future cannot be sent between threads safely
the trait `std::marker::Send` is not implemented for `dyn Future<Output = Result<(ledger_canister::Tokens,), (ic_kit::RejectionCode, std::string::String)>>`
required for the cast to the object type `dyn Future<Output = Result<u64, std::string::String>> + std::marker::Send`rustc
[principal_ext.rs(171, 22): ]()future is not `Send` as it awaits another future which is not `Send`
```

Reply:
It's still ok, for you don't need to use ic_kit in here. Just leave ic_cdk api call as is. When testing, this code will never be executed because of the logic I described above.

Reply:
Yep, seem to be working. Thanks. To avoid further confusion we need to document it somewhere though.